ruler/ruler.py

```diff
@@ -1,177 +1,265 @@
 import numpy as np
+from typing import Iterable, Optional, Tuple
 
-solid_indicator = ['solid','solids','black','s','b','true','1',True,1]
-void_indicator = ['invert','void','white','v','w','false','0',True,0]
-minimum_indicator = ['minimum','minimal','min']
-both_indicator = ['both','two','2',2]
-
-def ruler_initialize(arr,phys_size,margin_size):
-    if isinstance(phys_size,np.ndarray):
-        phys_size = phys_size[np.squeeze(phys_size).nonzero()] # physical size
-    elif isinstance(phys_size,list) or isinstance(phys_size,tuple):
-        phys_size = np.array(phys_size)[np.squeeze(np.array(phys_size)).nonzero()]
-    elif isinstance(phys_size,float) or isinstance(phys_size,int):
-        phys_size = np.array([phys_size])
-    else:
-        AssertionError("Invalid format of the physical size.")
+solid_indicator = ['solid', 'solids', 'black', 's', 'b', 'true', '1', True, 1]
+void_indicator = ['invert', 'void', 'white', 'v', 'w', 'false', '0', True, 0]
+minimum_indicator = ['minimum', 'minimal', 'min']
+both_indicator = ['both', 'two', '2', 2]
 
-    dims = len(phys_size) # dimension, should be 1, 2 or 3
-    assert np.squeeze(arr).ndim == dims, 'The physical size and the dimension of the input array do not match.'
 
+def minimum_length(
+        arr: Iterable[np.ndarray],
+        phys_size: Iterable[np.ndarray],
```

Comment:
> Does it make sense for these to be np.ndarrays (which, in general, can have any number of dimensions / shape)?

I think it makes sense for the input design pattern `arr` to be `np.ndarray`, but `phys_size` does not need to be `np.ndarray`. The code can accept `phys_size` as an array, a list, a tuple, or simply a number (for 1d design pattern).

> If I understand these parameters correctly, phys_size specifies the extent in physical units of the design arr along each of its axes?

Yes, this is correct.

> If so, could we just type this as Tuple[float, ...] or List[float] and validate that len(phys_size) == arr.ndim?

The argument `phys_size` can have different types as mentioned above. If we have to choose one of the types for type hints, why is `Tuple` or `List` preferred over `Iterable[np.ndarray]` (apart from appearing more straightforward)?

> Similar comment for margin_size. If I understand correctly, it is meant to be interpreted similarly to a padding, but is instead a sort of de-padding.

Yes, this is correct.

> I would suggest a type similar to the pad_width in numpy's [pad operation](https://numpy.org/doc/stable/reference/generated/numpy.pad.html) like List[Tuple[float, float]] where the quantities are [(before_1, after_1), ... (before_N, after_N)] for each axis.

This will be adopted in the next version.

Reply:
> The code can accept phys_size as an array, a list, a tuple, or simply a number (for 1d design pattern).

Supporting so many possibilities places a much higher burden on validating for correctness of the inputs. In particular, `np.ndarray` is a completely unconstrained type and seems quite awkward (IMO) for specifying the extent and border "throw away" region.

Constraining the API for `phys_size` and `margin_size` to be either a tuple or a list whose length matches `arr.ndim` seems reasonable and has precident in various numpy APIs.

Reply:
Agreed that we should keep things as simple as possible. To that end, it may be worthwhile removing the `phys_size` here and simply compute the minimum feature size in units of pixels. A separate API with a physical size could then be created by wrapping this simpler function.

Reply:
> To that end, it may be worthwhile removing the phys_size here and simply compute the minimum feature size in units of pixels. A separate API with a physical size could then be created by wrapping this simpler function.

The resolution along different dimensions may not always be the same. For example, in a 2d design pattern, the resolution along the x direction may not be the same as the resolution along the y axis, although we have not encountered this situation in this project. Such anisotropy makes it infeasible to decouple the two steps.